File: Game-Engine/src/engine/io/Window.java
package engine.io;

import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.glfw.GLFWVidMode;

import org.lwjgl.opengl.GL;
import org.lwjgl.opengl.GL11;

import static org.lwjgl.glfw.GLFW.*;

import engine.math.Vector4f;

public class Window {
    private long handle;
    private int width;
    private int height;
    private String title;
    private int frames;
    private static long time;
    private Vector4f background = new Vector4f(0,0,0, 1);

    public Window(int width, int height, String title){
        this.width = width;
        this.height = height;
        this.title = title;
        create();
    }

    private void create(){
        GLFWErrorCallback.createPrint(System.err).set();

        if (!glfwInit())
            throw new IllegalStateException("Unable to initialize GLFW");

        handle = glfwCreateWindow(width, height, title, 0, 0);

        if (handle == 0)
            throw new RuntimeException("Failed to create the GLFW window");

        // Get the resolution of the primary monitor
        GLFWVidMode vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());

        glfwSetWindowPos(handle,
                (vidmode.width() - width) / 2,
                (vidmode.height() - height) / 2); // Center the window
        glfwMakeContextCurrent(handle); // Make the OpenGL context current

        GL.createCapabilities();
        GL11.glEnable(GL11.GL_DEPTH_TEST);

        glfwShowWindow(handle); // Make the window visible
        glfwSwapInterval(1); // Enable v-sync

        time = System.currentTimeMillis();
    }

    public void update(){

        GL11.glClearColor(background.getX(), background.getY(), background.getZ(), background.getW());
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT);
        glfwPollEvents();

        frames++;
        // if more than a second has passed
        if (System.currentTimeMillis() > time + 1000) {
            glfwSetWindowTitle(handle, title + " | FPS: " + frames);
            time = System.currentTimeMillis();
            frames = 0;
        }
    }

    public void destroy() {
        glfwWindowShouldClose(handle);
        glfwDestroyWindow(handle);
        glfwTerminate();
    }

    public void swapBuffers(){
        glfwSwapBuffers(handle);
    }

    public boolean shouldClose(){
        return glfwWindowShouldClose(handle);
    }

    public void setBackgroundColor(float r, float g, float b, float a){
        background.setX(r);
        background.setY(g);
        background.setZ(b);
        background.setW(a);
    }

}

File: CPP-Game-Engine/CPP-Game-Engine/src/io/Window.cpp
#include "Window.h"

namespace sam_engine { namespace io {

	void window_resize(GLFWwindow* window, int width, int height);

	Window::Window(const char* name, int width, int height) {
		m_Name = name;
		m_Width = width;
		m_Height = height;
		m_Input = Input();
		if(!init())
			glfwTerminate();
	}

	Window::~Window() {
		
		glfwTerminate();

	}

	bool Window::init() {

		if (!glfwInit()) {
			std::cout << "Could not initialize GLFW" << std::endl;
			return false;
		}

		m_Window = glfwCreateWindow(m_Width, m_Height, m_Name, NULL, NULL);

		if (!m_Window) {
			std::cout << "Failed to create window" << std::endl;
			return false;
		}

		glfwMakeContextCurrent(m_Window);
		glfwSetWindowSizeCallback(m_Window, window_resize);

		if (glewInit() != GLEW_OK) {
			std::cout << "Could not initialize GLEW!" << std::endl;
			return false;
		}

		glfwSetKeyCallback(m_Window, this->m_Input.key_callback);
		glfwSetMouseButtonCallback(m_Window, this->m_Input.mouse_button_callback);
		glfwSetCursorPosCallback(m_Window, this->m_Input.cursor_pos_callback);

		return true;
		
	}

	void Window::clear() const {

		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

	}
	
	void Window::update() {

		GLenum error = glGetError();
		if (error != GL_NO_ERROR)
			std::cout << "OpenGL Error: " << error << std::endl;

		glfwSwapBuffers(m_Window);
		glfwPollEvents();

	}

	bool Window::closed() {
		return glfwWindowShouldClose(m_Window);
	}

	void Window::setBackground(float r, float g, float b) {
	
		glClearColor(r, g, b, 1.0f);
	
	}

	void window_resize(GLFWwindow* window, int width, int height) {
		glViewport(0, 0, width, height);
	}

}}
